Add vfs quota support

sargun opened the pull request:
**- What I did**
Added project quota support to the VFS graphdriver, and tests to ensure they work. 

It relies on https://github.com/moby/moby/pull/35177, which adds the quota detection mechanism, and a form error to determine whether to short-circuit quota-related code (or not).

It addresses this issue: https://github.com/moby/moby/issues/34702

**- How I did it**
I reused the project quota code that came from overlay2. On initialization time, the VFS driver (on Linux) detects if quota is supported. If quota is supported, then it enables the MaxSize option, so new containers can be created with an assigned project quota.

**- How to verify it**
There are tests as part of the changeset. If you run the tests when your `TMPDIR` environment variable points to an XFS mount mounted with `prjquota` option, the quota check will run automatically, otherwise, it'll be skipped.

**- Description for the changelog**
Add project quota support to VFS graphdriver

AkihiroSuda commented:
what motivates you to use vfs?

sargun commented:
@AkihiroSuda "It just works" (mostly). My containers run from 2 minutes to 2 months. The median runtime of my containers is 4-5 minutes. Setting up VFS, with the improvements that come from https://github.com/moby/moby/pull/34670 has an unoptimized mean time of ~45s for copying ubuntu:latest. We can optimize that down to <10s. 

Never having to worry about overlayfs, or such, there's some appeal there.

sargun commented:
@dmcgowan We discussed this change earlier.

thaJeztah commented on daemon/graphdriver/graphtest/graphtest_unix.go:
```diff
@@ -12,6 +12,8 @@ import (
 	"unsafe"
 
+	"github.com/docker/docker/daemon/graphdriver/quota"
+
```
nit: can you remove this blank line?

thaJeztah commented:
/cc @kolyshkin as well

Commit:
Add quota support to VFS graphdriver

This patch adds the capability for the VFS graphdriver to use
XFS project quotas. It reuses the existing quota management
code that was created by overlay2 on XFS.

It doesn't rely on a filesystem whitelist, but instead
the quota-capability detection code.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

sargun commented:
Rebased.

sargun commented:
This should be ready to go in.

kolyshkin commented on daemon/graphdriver/graphtest/graphtest_unix.go:
```diff
@@ -310,15 +311,17 @@ func writeRandomFile(path string, size uint64) error {
 	createOpts.StorageOpt = make(map[string]string, 1)
 	createOpts.StorageOpt["size"] = "50M"
-	if err := driver.Create("zfsTest", "Base", createOpts); err != nil {
+	if err := driver.CreateReadWrite("zfsTest", "Base", createOpts); err == quota.ErrQuotaNotSupported && !required {
```
Perhaps it makes sense to rename `"zfsTest"` to `drivername + "Test"` as it's no longer zfs-specific. Or just `"quotaTest"`...

kolyshkin commented on daemon/graphdriver/graphtest/graphtest_unix.go:
```diff
@@ -329,8 +332,14 @@ func DriverTestSetQuota(t *testing.T, drivername string) {
+	}
+
+	// Try to write a file bigger than quota, and ensure it does not work
+	if pathError, ok := err.(*os.PathError); ok && pathError.Err != unix.EDQUOT && pathError.Err != unix.ENOSPC {
```
1. Looks like you're missing a line here? Something like
```golang
+	err = writeRandomFile(path.Join(mountPath.Path(), "file"), quota*2)
```

2. Since you have already exhausted half the quota by doing a write above, you might either go with a smaller value than `quota*2`, or remove the file written above.

kolyshkin commented on daemon/graphdriver/graphtest/graphtest_unix.go:
```diff
@@ -329,8 +332,14 @@ func DriverTestSetQuota(t *testing.T, drivername string) {
+
+	// Try to write a file bigger than quota, and ensure it does not work
+	if pathError, ok := err.(*os.PathError); ok && pathError.Err != unix.EDQUOT && pathError.Err != unix.ENOSPC {
+		t.Fatalf("expect write() to fail with %v or %v, got %v", unix.EDQUOT, unix.ENOSPC, err)
```
Perhaps you want to print ``pathError.Err`` here, not ``err``, as this is what you use in comparisons?

kolyshkin commented on daemon/graphdriver/vfs/driver.go:
```diff
@@ -41,8 +48,10 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
-	home       string
-	idMappings *idtools.IDMappings
+	driverQuota
+	quotaSupported bool
```
I think this can be done in a simple manner, without an extra `bool` field. The idea is to add `isQuotaSupported()` method. In `quota_linux.go`, it is just `return driver.quotaCtl != nil`, in `quota_unsupported.go`, return false straight away. To my mind, it will be cleaner/simpler.

sargun commented:
Updated based on review.

cpuguy83 commented on daemon/graphdriver/vfs/quota_unsupported.go:
```diff
@@ -0,0 +1,18 @@
+}
+
+func (d *Driver) setupQuota(dir string, size uint64) error {
+	panic("Platform does not support quota; this should never be called")
```
Panic seems a little harsh here. Can we just return a not supported error?

sargun replied:
This should never be called. If this is called, it means something has gone wrong, or the code was changed in an unexpected way.

justincormack commented:
ping @cpuguy83

justincormack commented on daemon/graphdriver/graphtest/graphtest_unix.go:
```diff
@@ -310,27 +311,42 @@ func writeRandomFile(path string, size uint64) error {
+	}
+	defer os.Remove(path.Join(mountPath.Path(), "smallfile"))
+
+	// Try to write a file bigger than quota, given that we've already written a file larger than quota, and ensure it does not work
```
this comment isn't right, we have written a file of half quota at this point?

kolyshkin reviewed (approved):
LGTM

cpuguy83 commented on daemon/graphdriver/quota/errors.go:
```diff
@@ -0,0 +1,6 @@
+import "errors"
+
+// ErrQuotaNotSupported indicates if were found the FS didn't have projects quotas available
+var ErrQuotaNotSupported = errors.New("Filesystem does not support, or has not enabled quotas")
```
Can we make sure this error implements one of api/errdefs?

sargun replied:
Which one do you think it should be?

sargun replied:
ErrNotImplemented?

sargun replied:
Or ErrUnavailable

cpuguy83 replied:
NotImplemented sounds good.

sargun commented:
@cpuguy83 Anything else?

cpuguy83 reviewed (approved):
LGTM
